[butteredyakiimo] iP

butteredyakiimo opened the pull request:
# BUTTER
> "Tracking your tasks will be as smooth as butter. :butter:"

BUTTER helps you remember what you need to do. It's,
- text-based
- easy to use 
- ~FAST~ _SUPER_ FAST to use

All you need to do is, 
1. download it from [here.](https://github.com/butteredyakiimo/ip/releases)
2. double click it.
3. add your tasks.
4. let it manage your tasks for you!

And it is **FREE**!

## Features:
- [x] Managing tasks
- [x] Managing deadlines 
- [x] Managing events
- [x] Searching for tasks
- [x] Reminders (coming soon)

If you are a Java programmer, you can use it to practise Java too. Here's the `main` method: 

```java
public static void main(String[] args) {
        new Duke("./data/duke.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

butteredyakiimo committed:
Change name of chatbot and add greetings

butteredyakiimo committed:
Add user input, chatbot echos command, exits with bye

butteredyakiimo committed:
Store user input, list out all tasks

butteredyakiimo committed:
Add a Task class, users can mark tasks as done

butteredyakiimo committed:
Add commands todo, deadline and event, created these classes too

butteredyakiimo committed:
Add TextUiTesting, made some fields private

butteredyakiimo committed:
Handle errors, such as invalid and incomplete commands

butteredyakiimo committed:
Add delete function, fix invalid task id for mark, unmark and delete

butteredyakiimo committed:
Update changes to Duke.java, include fixes mark, unmark and delete

butteredyakiimo committed:
Add exceptions

butteredyakiimo committed:
Tasks are saved in another file, updates when list changes

butteredyakiimo committed:
Merge branch 'branch-Level-7'

butteredyakiimo committed:
Change date to LocalDateTime for deadline

butteredyakiimo committed:
Dates in Events are now LocalDateTime

butteredyakiimo committed:
Merge branch 'branch-Level-8'

butteredyakiimo committed:
Extract out Ui, Storage, Parser and TaskList classes

butteredyakiimo committed:
Add new exceptions, improve abtraction of ui class

butteredyakiimo committed:
Merge branch 'branch-A-MoreOOP'

butteredyakiimo committed:
Create packages: duke, exceptions and tasks

butteredyakiimo committed:
Merge branch 'branch-A-Packages'

butteredyakiimo committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552' into A-Gradle

butteredyakiimo committed:
Set up gradle to build and run Duke

butteredyakiimo committed:
Merge branch 'A-Gradle'

butteredyakiimo committed:
Add tests for all classes under tasks package

butteredyakiimo committed:
Add tests for all classes under tasks package

butteredyakiimo committed:
Merge branch 'branch-A-JUnit'

butteredyakiimo committed:
Create JAR file

butteredyakiimo committed:
Merge branch 'branch-A-JAR'

butteredyakiimo committed:
No exceptions are caught in Duke class, handled instead by parser

butteredyakiimo committed:
Merge branch 'branch-Fix-Exceptions'

butteredyakiimo committed:
Add Java docs for all classes, except exceptions

butteredyakiimo committed:
Edit code to adhere to coding standards

derekjxtan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,190 @@
+            } else {
+                //0 for unmarked, any other number for marked
+                ToDo toDo = new ToDo(0, inputArr[1]);
+                tasks.addTask(toDo, storage);
```
Perhaps acting on the inputs can be put in another class? IMO the name parser doesn't really indicate that this behaviour is expected.

derekjxtan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,121 @@
+        return dataFile;
+    }
+
+    public ArrayList<Task> loadTasks() {
```
personally i feel that Storage should only be responsible for reading what is stored in the file. Maybe the logic for converting the text to the Task could be moved to parser or somewhere else relevant. What do you think?

derekjxtan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    //this method should be called by the tasklist class ONLY!!! because we want to
+    //keep the tasklist private
```
I think rather than adding a comment, it would be better for TaskList to generate a ArrayList of strings for the Storage class to write to file. This ensures that the the ArrayList of Tasks would not be passed to an external class. This make the Storage class more reusable as well.

derekjxtan reviewed (commented):
LGTM. Overall great use of OOP principles and code quality is good!

butteredyakiimo committed:
Add find feature

butteredyakiimo committed:
Merge branch 'branch-A-JavaDoc'

butteredyakiimo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/tasks/Task.java
#	src/test/java/duke/tasks/TaskTest.java

butteredyakiimo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/duke.txt
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/tasks/Task.java
#	src/test/java/duke/tasks/TaskTest.java

butteredyakiimo committed:
Fix renaming of ID to id in exceptions and methods

ZD292 commented on .gitignore:
```diff
@@ -15,3 +15,5 @@ bin/
 /text-ui-test/ACTUAL.TXT
 text-ui-test/EXPECTED-UNIX.TXT
+
+/src/main/java/*.class
```
Looks good to me 😄

butteredyakiimo committed:
Add more JavaDocs

butteredyakiimo committed:
Merge branch 'branch-Fix-Merge-Bugs'

butteredyakiimo committed:
Fix cosmetic issues for javadocs

antonTan96 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+
+        while (scanner.hasNextLine()) {
+            String input = scanner.nextLine();
+            if (parser.isGoodbye(input)) {
+                break;
+            } else {
+                parser.parseInput(input);
+            }
```
The way you check for a terminating condition is quite intuitive by checking whether the input is "bye" or not. However, this involves throwing the input into the parser twice. Would it be better to let the parser handle the case "bye" in `parser.parseInput(input)` ?

antonTan96 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,318 @@
+     * @return true if s can be parsed as an integer.
+     */
+    //credit: https://www.freecodecamp.org/news/java-string-to-int-how-to-convert-a-string-to-an-integer/
+    private static boolean isNumber(String s) {
+        return s != null && s.matches("[0-9.]+");
+    }
```
I like how regex is used here :D

antonTan96 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,318 @@
+     */
+    public void parseMark(String input) throws NoTaskIdException, InvalidTaskIdException {
+        String[] inputArr = input.split(" ", 2);
+        if (inputArr.length == 1) {
+            throw new NoTaskIdException();
+        } else {
+            String strIndex = inputArr[1];
+            if (isNumber(strIndex)) {
+                int index = Integer.parseInt(strIndex) - 1; //because index starts from 1
+                if (tasks.isValidTaskId(index)) {
+                    tasks.markTask(index);
+                }
+            } else {
+                throw new InvalidTaskIdException();
+            }
+        }
```
What happens if `tasks.isValidTaskId(index)` returns false? Could InvalidTaskIdException be thrown? I would write this part like this:
```
        if (inputArr.length == 1) {
            throw new NoTaskIdException();
        } else {
            String strIndex = inputArr[1];
            if (isNumber(strIndex) && tasks.isValidTaskId(Integer.parseInt(strIndex) - 1)) {
                int index = Integer.parseInt(strIndex) - 1; //because index starts from 1
                
                tasks.markTask(index);
                
            } else {
                throw new InvalidTaskIdException();
            }
        }
```

antonTan96 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,318 @@
+     */
+    public void parseUnMark(String input) throws NoTaskIdException, InvalidTaskIdException {
+        String[] inputs = input.split(" ", 2);
+        if (inputs.length == 1) {
+            throw new NoTaskIdException();
+        } else {
+            String strIndex = inputs[1];
+            if (isNumber(strIndex)) {
+                int index = Integer.parseInt(strIndex) - 1; //because index starts from 1
+                if (tasks.isValidTaskId(index)) {
+                    tasks.unMarkTask(index);
+                }
+            } else {
+                throw new InvalidTaskIdException();
+            }
+        }
```
Same question as mentioned above

antonTan96 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,318 @@
+     */
+    public void parseDeadline(String input) throws NoDescException, InvalidDeadlineException {
+        String[] inputs = input.split(" ", 2);
+        if (inputs.length == 1) {
+            throw new NoDescException();
+        } else {
+            String afterCommand = inputs[1];
+            String[] details = afterCommand.split(" /by ", 2);
+
+            if (details.length < 2) {
+                throw new InvalidDeadlineException();
+            } else {
+                String desc = details[0];
+                String date = details[1];
+
+                if (desc.isEmpty()) {
+                    throw new NoDescException();
+                } else {
+                    try {
+                        LocalDateTime dateTime = convertToDateTime(date);
+                        Deadline deadline = new Deadline(0, desc, dateTime);
+                        tasks.addTask(deadline);
+                    } catch (DateTimeParseException e) {
+                        ui.showInvalidDateFormat();
+                    }
+                }
+            }
+        }
```
This part is quite heavily nested. Could you try to refactor the code to make it more flatter?

antonTan96 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,318 @@
+            throw new NoDescException();
+        } else {
+            String afterCommand = inputs[1];
+            String[] details = afterCommand.split(" /from ", 2);
+
+            if (details[0].isBlank()) {
+                throw new NoDescException();
+            } else if (details.length == 1) {
+                throw new InvalidEventException();
+            } else {
+                String task = details[0];
+                String start = details[1].split(" /to ", 2)[0];
+
+                if (start.isBlank()) {
+                    throw new NoStartException();
+                } else {
+                    String[] endDetails = afterCommand.split(" /to ", 2);
+
+                    if (endDetails.length == 1) { //no end date added
+                        throw new NoEndException();
+                    } else {
+                        String end = endDetails[1];
+
+                        if (end.isBlank()) {
+                            throw new NoEndException();
+                        } else {
+                            try {
+                                LocalDateTime startDateTime = convertToDateTime(start);
+                                LocalDateTime endDateTime = convertToDateTime(end);
+                                Event event = new Event(0, task, startDateTime, endDateTime);
+                                tasks.addTask(event);
+                            } catch (DateTimeParseException e) {
+                                ui.showInvalidDateFormat();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
```
Same question with this method.

antonTan96 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,156 @@
+                dataFile = createDataFile();
+            }
+
+            //create a FileWriter object to write to file. Note that this overwrites the existing data!
```
I like this warning for i have fallen for this so many times D:
I still don't know how FileWriters work

antonTan96 reviewed (commented):
LGTM! There could be some parts that can be optimized but it's a great effort overall

butteredyakiimo committed:
Fix nested code

butteredyakiimo committed:
Add more unit tests

butteredyakiimo committed:
Fix style violations

butteredyakiimo committed:
Merge branch 'branch-A-CheckStyle'

butteredyakiimo committed:
Add GUI

butteredyakiimo committed:
Change layout

butteredyakiimo committed:
Merge branch 'branch-Level-10'

butteredyakiimo committed:
Add configurations to build.gradle to create jar file

butteredyakiimo committed:
Add assertions to some classes.

Assertions help to detect possible bugs, as they
verify certain assumptions that the code makes.
Gradle run is also configured to enable assertions.

butteredyakiimo committed:
Modify code to adhere to code quality standards.

Some methods are extremely long, and have more than one
level of abstraction. Magic numbers are used to describe
the task status.

Let's make more methods to adhere to SLAP and avoid nested code,
and also create a new Enum to make code easier to understand.

butteredyakiimo committed:
Merge pull request #2 from butteredyakiimo/branch-A-Assertions

Add assertions to some classes

butteredyakiimo committed:
Merge branch 'master' into branch-A-CodeQuality

butteredyakiimo committed:
Merge pull request #1 from butteredyakiimo/branch-A-CodeQuality

Modify code to adhere to code quality standards

butteredyakiimo committed:
Add a tag to the list command

Currently, users only have a list commmand to list all the tasks.

To create reminders for upcoming tasks, users need to be able to view
the tasks that are due before/within the time period.

A flag can be used at the end of the list command of the format:
'list flag'. The time period for both deadline and event starts at the
start of the day at 00:00. The end is determined by the flag indicated
by the user.

Let's add the following flags:
* 'today' : lists all tasks that in the period today 00:00 to tmr 00:00
* 'week' : lists all tasks that are in the period today 00:00 to one
week later 00:00
* 'fort': lists all tasks that are in the period today 00:00 to 2 weeks
later 00:00

butteredyakiimo committed:
Add a tag to the list command

Currently, users only have a list commmand to list all the tasks.

To create reminders for upcoming tasks, users need to be able to view
the tasks that are due before/within the time period.

A flag can be used at the end of the list command of the format:
'list flag'. The time period for both deadline and event starts at the
start of the day at 00:00. The end is determined by the flag indicated
by the user.

Let's add the following flags:
* 'today' : lists all tasks that in the period today 00:00 to tmr 00:00
* 'week' : lists all tasks that are in the period today 00:00 to one
week later 00:00
* 'fort': lists all tasks that are in the period today 00:00 to 2 weeks
later 00:00

butteredyakiimo committed:
Merge branch 'branch-Reminders'

butteredyakiimo committed:
Fix style

butteredyakiimo committed:
Add background color for response dialog label

butteredyakiimo committed:
Set title of stage to BUTTER

butteredyakiimo committed:
Merge branch 'branch-A-BetterGui'

butteredyakiimo committed:
Add user guide

butteredyakiimo committed:
Fix paragraph issue in user guide

butteredyakiimo committed:
Modify code to meet ip standards

butteredyakiimo committed:
Modify list [FLAG] to include todos
